File: humble_catalog/stats.py
import sys

from humble_catalog import db

# Each gap is "a column is falsy" (NULL, missing, or empty string),
# counted across ALL items -- consistent with the viewer's `unrated`
# filter, which also ignores enrichment status. flag_key matches the
# viewer's #f-flag values so the CLI and the viewer name the same gaps.
GAPS = (
    ("Unrated", "unrated", "my_rating"),
    ("No cover", "nocover", "cover_path"),
    ("No source URL", "nourl", "source_url"),
)

# Fixed vocabularies, each paired with its display label. Row order comes
# from these tuples and never from the counts, so the panel does not
# reshuffle as the library changes. TYPES matches index.html's #f-type
# options; READ_STATUSES is the lifecycle order the viewer also sorts by;
# ENRICHMENT is the enrichment.status vocabulary.
TYPES = (("ebook", "E-books"), ("audiobook", "Audiobooks"),
         ("comic", "Comics"), ("music", "Music/Soundtracks"),
         ("android", "Android apps"))
READ_STATUSES = (("want_to_read", "Want to read"), ("unread", "Unread"),
                 ("reading", "Reading"), ("read", "Read"), ("dnf", "DNF"))
ENRICHMENT = (("matched", "Matched"), ("low_confidence", "Low confidence"),
              ("unmatched", "Unmatched"), ("pending", "Pending"))


def _tally(items, field, vocab, default=None):
    """Count `field` over `items` against a fixed (value, label) vocab.

    A value outside the vocab is counted nowhere rather than inventing a
    row, so a section need not sum to the total. `default` stands in for a
    missing key, which is how read_status's NOT NULL DEFAULT is honoured
    for a partial payload from an older server.
    """
    counts = {value: 0 for value, _label in vocab}
    for item in items:
        value = item.get(field) or default
        if value in counts:
            counts[value] += 1
    return [(label, counts[value]) for value, label in vocab]


def _by_type(items):
    return _tally(items, "type", TYPES)


def _by_rating(items):
    # 1..5 only: unrated is a gap, and a row in both places would break the
    # one-number-one-place rule the panel exists to restore.
    return [(f"★{n}", sum(1 for i in items if i.get("my_rating") == n))
            for n in range(1, 6)]


def _by_read_status(items):
    return _tally(items, "read_status", READ_STATUSES, default="unread")


def _by_enrichment(items):
    return _tally(items, "status", ENRICHMENT)


def _by_gap(items):
    return [(label, sum(1 for i in items if not i.get(field)))
            for label, _flag, field in GAPS]


def _by_genre(items):
    # The one section ordered by its data: biggest first, ties broken
    # alphabetically so the order is total and stable across runs. Every
    # tag is returned; showing only a top N is the viewer's business.
    counts = {}
    for item in items:
        for tag in item.get("genre") or ():
            counts[tag] = counts.get(tag, 0) + 1
    return sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))


# The report's order. Keys are the viewer's contract: each maps to the
# filter a row of that section applies (see SECTION_FILTERS in app.js).
SECTIONS = (
    ("type", "By type", _by_type),
    ("rating", "Ratings", _by_rating),
    ("status", "Reading status", _by_read_status),
    ("enrichment", "Enrichment", _by_enrichment),
    ("gaps", "Gaps", _by_gap),
    ("genre", "Genres", _by_genre),
)


def report(items):
    """Every section's counts over `items` (a db.fetch_items list).

    Returns (sections, total): sections is [(key, label, rows), ...] in
    SECTIONS order, where rows is [(label, count), ...]; total is
    len(items). Pure -- no connection, no I/O -- so the CLI and the web
    route count over the same per-item view fetch_items feeds /api/items
    and the CSV export, which is what keeps them from drifting.
    """
    return ([(key, label, rows_fn(items)) for key, label, rows_fn in SECTIONS],
            len(items))


def console_safe(text, encoding):
    """Degrade `text` to what a console using `encoding` can actually print.

    The default Windows console is cp1252, which has no ★, and genre tags
    are arbitrary user data that may hold anything. The labels are shared
    with the web panel, where the star is the right character, so the CLI
    degrades them at its own boundary rather than the vocabulary being
    dulled for every surface. ★ becomes * because "?5" reads as a
    mistake; anything else falls back to the codec's replacement.
    """
    try:
        text.encode(encoding)
        return text
    except UnicodeEncodeError:
        text = text.replace("★", "*")
        return text.encode(encoding, errors="replace").decode(encoding)


def run(conn):
    sections, total = report(db.fetch_items(conn))
    width = len(str(total))
    encoding = getattr(sys.stdout, "encoding", None) or "utf-8"
    for _key, label, rows in sections:
        print(f"\n{label}")
        for row_label, count in rows:
            print(f"{count:>{width}}  {console_safe(row_label, encoding)}")
    print(f"\n{total:>{width}}  items total")

File: humble_catalog/webapp/static/stats.js
// The statistics panel's counting, in the browser.
//
// This is a second implementation of stats.py's counts, which is a drift
// risk the server-side arrangement was chosen to avoid. What holds the
// two together now is a test rather than an arrangement: one fixture is
// counted by both and asserted equal field for field (see
// test_the_browser_counts_agree_with_stats_py). The same invariant the
// probe battery pins for /api/stats, moved to the pair that can disagree.
//
// The counting had to move because the panel describes the rows the
// table is showing, and the browser is the only place that knows which
// rows those are: the filters never reach the server, and /api/stats
// documents that nothing about the library may travel in a query string.

// Gaps are the one vocabulary the viewer did not already hold -- it keeps
// each gap's FILTER flag, not the column the gap is in. Order matches
// stats.py's GAPS.
const GAP_FIELDS = {"Unrated": "my_rating", "No cover": "cover_path",
                    "No source URL": "source_url"};

// Count `field` against a label -> value vocabulary. Those vocabularies
// live in catalog.js, already in stats.py's order because the
// overview-to-filter jump needs them, so the viewer keeps one list and
// not two. Object key order is insertion order for string keys, which is
// what makes them usable as ordered vocabularies.
//
// A value outside the vocabulary is counted nowhere rather than
// inventing a row, so a section need not sum to the total; `dflt` stands
// in for a missing key, which is how read_status's NOT NULL DEFAULT is
// honoured for a partial payload. Both match stats.py's _tally.
function statTally(items, field, labelsToValues, dflt) {
  const counts = new Map(Object.values(labelsToValues).map((v) => [v, 0]));
  for (const item of items) {
    const value = item[field] || dflt;
    if (counts.has(value)) counts.set(value, counts.get(value) + 1);
  }
  return Object.entries(labelsToValues)
    .map(([label, value]) => ({label, count: counts.get(value)}));
}

// 1..5 only: unrated is a gap, and a row in both places would break the
// one-number-one-place rule the panel exists to restore.
const statsByRating = (items) =>
  [1, 2, 3, 4, 5].map((n) => ({label: `★${n}`,
    count: items.filter((i) => i.my_rating === n).length}));

const statsByGap = (items) =>
  Object.entries(GAP_FIELDS).map(([label, field]) =>
    ({label, count: items.filter((i) => !i[field]).length}));

// The one section ordered by its data: biggest first, ties broken
// alphabetically so the order is total and stable. Comparing with < is
// by UTF-16 code unit, which agrees with Python's code-point ordering
// for everything below the astral planes.
function statsByGenre(items) {
  const counts = new Map();
  for (const item of items)
    for (const tag of item.genre || [])
      counts.set(tag, (counts.get(tag) || 0) + 1);
  return [...counts]
    .sort((a, b) => b[1] - a[1] || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0))
    .map(([label, count]) => ({label, count}));
}

// Order and labels mirror stats.py's SECTIONS; the keys are the contract
// SECTION_FILTERS reads back when a row is clicked.
const STATS_SECTIONS = [
  ["type", "By type", (i) => statTally(i, "type", TYPE_VALUES)],
  ["rating", "Ratings", statsByRating],
  ["status", "Reading status",
   (i) => statTally(i, "read_status", STATUS_VALUES, "unread")],
  ["enrichment", "Enrichment", (i) => statTally(i, "status", ENRICHMENT_VALUES)],
  ["gaps", "Gaps", statsByGap],
  ["genre", "Genres", statsByGenre],
];

// The shape /api/stats returned, so renderStats and the panel's tests are
// untouched by where the numbers come from.
function statsReport(items) {
  return {total: items.length,
          sections: STATS_SECTIONS.map(([key, label, rows]) =>
            ({key, label, rows: rows(items)}))};
}
